org.eclipse.xtext.java/src/org/eclipse/xtext/java/resource/InMemoryClassLoader.java

```
@@ -0,0 +1,107 @@
+
+	public InMemoryClassLoader(Map<String, byte[]> classMap, ClassLoader parent) {
+		super(parent);
+		this.classMap = classMap;
+	}
+
+	@Override
+	protected Class<?> findClass(String name) throws ClassNotFoundException {
+		byte[] bytes = classMap.get(name);
+		if (bytes == null) {
+			return super.findClass(name);
+		} else {
+			return defineClass(name, bytes, 0, bytes.length);
+		}
+	}
+
+	@Override
+	public URL getResource(String path) {
+		try {
+			if (path.endsWith(".class")) {
+				String className = pathToClassName(path);
+				byte[] bytes = classMap.get(className);
+				if (bytes != null) {
+					return new URL("in-memory", null, -1, path, new URLStreamHandler() {
+						@Override
+						protected URLConnection openConnection(URL it) throws IOException {
+							return new URLConnection(it) {
+								@Override
+								public void connect() {
+								}
+
+								@Override
+								public InputStream getInputStream() {
+									return new ByteArrayInputStream(bytes);
+								}
+							};
+						}
+					});
+				}
+				return super.getResource(path);
+			}
+			return null;
+		} catch (MalformedURLException e) {
+			throw Exceptions.sneakyThrow(e);
+		}
+	}
+
+	@Override
+	protected URL findResource(String path) {
+		try {
+			if (path.endsWith(".class")) {
+				String className = pathToClassName(path);
+				byte[] bytes = classMap.get(className);
+				if (bytes != null) {
+					return new URL("in-memory", null, -1, path, new URLStreamHandler() {
+						@Override
+						protected URLConnection openConnection(URL it) throws IOException {
+							return new URLConnection(it) {
+								@Override
+								public void connect() {
+								}
+
+								@Override
+								public InputStream getInputStream() {
+									return new ByteArrayInputStream(bytes);
+								}
+							};
+						}
+					});
+				}
+			}
+			return super.findResource(path);
+		} catch (MalformedURLException e) {
+			throw Exceptions.sneakyThrow(e);
```

We don't like sneakies

---

i want to keep it here as i dont want to break. see above

---

what is your alternative in this case.
swallow the exception and do nothing?

---

Either return null (and log) or rethrow the error.

---

Retro’s is what sneaky does. The api of classloaddr is bs. I don’t want to do some fancy different exception handler right now

---

Okay